### PR TITLE
Fix invalid special casing of the unreachable! macro

### LINKED ISSUE #92137: unreachable!("{}") works on Rust 2021
The behavior of `unreachable` does not depend the edition of the crate that invoked the macro. That's fine for `todo!()` and `unimplemented!()` due to the way they are implemented, but in #92068, it was discovered that `unreachable!()` behaves subtly different, due to the way it has a special one-argument case. Looks like missed this when working on the 2021 panic macros. :(

### PR DESCRIPTION
**Pull request by Urgau**
This pull-request fix an invalid special casing of the `unreachable!` macro in the same way the `panic!` macro was solved, by adding two new internal only macros `unreachable_2015` and `unreachable_2021` edition dependent and turn `unreachable!` into a built-in macro that do dispatching. This logic is stolen from the `panic!` macro.

~~This pull-request also adds an internal feature `format_args_capture_non_literal` that allows capturing arguments from formatted string that expanded from macros. The original RFC #2795 mentioned this as a future possibility. This feature is [required](https://github.com/rust-lang/rust/issues/92137#issuecomment-1018630522) because of concatenation that needs to be done inside the macro:~~
```rust
$crate::concat!("internal error: entered unreachable code: ", $fmt)
```

**In summary** the new behavior for the `unreachable!` macro with this pr is:

Edition 2021:
```rust
let x = 5;
unreachable!("x is {x}");
```
```
internal error: entered unreachable code: x is 5
```

Edition <= 2018:
```rust
let x = 5;
unreachable!("x is {x}");
```
```
internal error: entered unreachable code: x is {x}
```

Also note that the change in this PR are **insta-stable** and **breaking changes** but this a considered as being a [bug](https://github.com/rust-lang/rust/issues/92137#issuecomment-998441613).
If someone could start a perf run and then a crater run this would be appreciated.

Fixes https://github.com/rust-lang/rust/issues/92137

### CODE REVIEW

File: src/test/ui/macros/unreachable-2021.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// edition:2021
```
Comment (camelid) [inline src/test/ui/macros/unreachable-2021.rs:1]: Could you rename this test to something like `unreachable-2021-format_args.rs` to make it clear what it tests?

Comment (camelid) [reply src/test/ui/macros/unreachable-2021.rs:1]: And could you also merge this test with the `unreachable-2015` one using [revisions](https://rustc-dev-guide.rust-lang.org/tests/compiletest.html#revisions)?

**Comment by camelid** [comment]
cc discussion about implementation strategy: https://github.com/rust-lang/rust/issues/92137#issuecomment-1018904497

**Comment by m-ou-se** [comment]
Thanks for working on this! I'll try to review this soon.

I'll have to note that for any edition-specific changes, we also need to add a migration lint to automatically upgrade code from older editions. (Just like we already have for `panic!()`.) Unfortuantely, because Rust 2021 is already stable, this will have to happen at the same time in this case.

> This pull-request also adds an internal feature `format_args_capture_non_literal` that allows capturing arguments from formatted string that expanded from macros. The original RFC #2795 mentioned this as a future possibility. This feature is [required](https://github.com/rust-lang/rust/issues/92137#issuecomment-1018630522) because of concatenation that needs to be done inside the macro:

Another option is to make that work using a nested format_args:

```diff
-        $crate::panic!($crate::concat!("internal error: entered unreachable code: ", $fmt), $($arg)*)
+        $crate::panic!("internal error: entered unreachable code: {}", $crate::format_args!($fmt, $($arg)*))
```

File: compiler/rustc_builtin_macros/src/unreachable.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
+use rustc_ast::ptr::P;
```
Comment (m-ou-se) [inline compiler/rustc_builtin_macros/src/unreachable.rs:1]: This file is almost an exact duplicate of `compiler/rustc_builtin_macros/src/panic.rs`. I'd prefer not to duplicate code like this, as it makes maintanance and history tracking more complicated. Instead, we could put both in the same file, and extract the common parts into a common function. E.g. something like:

```rust
pub fn expand_panic<'cx>(cx: &'cx mut ExtCtxt<'_>, sp: Span, tts: TokenStream) -> Box<dyn MacResult + 'cx> {
    let mac = if use_panic_2021(sp) { sym::panic_2021 } else { sym::panic_2015 };
    expand(mac, cx, sp, tts)
}

pub fn expand_unreachable<'cx>(cx: &'cx mut ExtCtxt<'_>, sp: Span, tts: TokenStream) -> Box<dyn MacResult + 'cx> {
    let mac = if use_panic_2021(sp) { sym::unreachable_2021 } else { sym::unreachable_2015 };
    expand(mac, cx, sp, tts)
}

fn expand(mac: Symbol, cx: &'cx mut ExtCtxt<'_>, sp: Span, tts: TokenStream) -> Box<dyn MacResult + 'cx> {
    ...
}
```

File: compiler/rustc_builtin_macros/src/unreachable.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+// This expands to either
+// - `$crate::panic::unreachable_2015!(...)` or
+// - `$crate::panic::unreachable_2021!(...)`
+// depending on the edition.
+//
+// `$crate` will refer to either the `std` or `core` crate depending on which
+// one we're expanding from.
```
Comment (m-ou-se) [inline compiler/rustc_builtin_macros/src/unreachable.rs:15]: This part is not true for `unreachable`, as `std::unreachable` is just a re-export of `core::unreachable`. So `$crate` will always be `core` here.

**Comment by m-ou-se** [comment]
> we also need to add a migration lint

I'd be happy to extend the `non_fmt_panics` lint to also cover unreachable. It's not trivial as `panic!(123)` and `unreachable!(123)` are a bit different: the first throws a `i32` as a `Box<dyn Any>`, while the latter just formats the `i32` with `"{}"`. But it shouldn't be too much work.

File: src/test/ui/macros/unreachable-2021.rs (hunk 1 above)
Comment (Urgau) [reply src/test/ui/macros/unreachable-2021.rs]: Done in https://github.com/rust-lang/rust/pull/93179/commits/d0e4240a1468b3b2e921324cbb7088850dacb4df

File: compiler/rustc_builtin_macros/src/unreachable.rs (hunk 2 above)
Comment (Urgau) [reply compiler/rustc_builtin_macros/src/unreachable.rs]: Done in https://github.com/rust-lang/rust/pull/93179/commits/ac1865fa02e4e2967bed87891ff79be30b9e593e

**Comment by Urgau** [comment]
I've updated the implementation to use a nested `format_args!` and reverted the introduction of `format_args_capture_non_literal`. I've also updated and improved the test using revisions as *camelid* [suggested](https://github.com/rust-lang/rust/pull/93179#discussion_r790090480).

> > we also need to add a migration lint
> 
> I'd be happy to extend the `non_fmt_panics` lint to also cover `unreachable`. It's not trivial as `panic!(123)` and `unreachable!(123)` are a bit different: the first throws a `i32` as a `Box<dyn Any>`, while the latter just formats the `i32` with `"{}"`. But it shouldn't be too much work.

I tried to expand the current `non_fmt_panics` but I'm stuck on the fact that the argument passed to `panic_str` is the result of `$crate::concat!("internal error: entered unreachable code: ", $fmt)` and I don't know how (if even possible) to get the argument passed to `concat` or `unreachable_2015`. Any suggestions ?

File: src/test/ui/macros/unreachable-format-args.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// ignore-emscripten no processes
```
Comment (camelid) [inline src/test/ui/macros/unreachable-format-args.rs:1]: Why are you ignoring emscripten in these tests?

**Comment by m-ou-se** [comment]
> > 
> 
> I tried to expand the current `non_fmt_panics` but I'm stuck on the fact that the argument passed to `panic_str` is the result of `$crate::concat!("internal error: entered unreachable code: ", $fmt)` and I don't know how (if even possible) to get the argument passed to `concat` or `unreachable_2015`. Any suggestions ?

Maybe you could make the second case of `unreachable_2015` expand to a function call instead: `unreachable_display(&$msg)` and then define a `fn unreachable_display(_: &impl Display) -> !;` to do the panic. Then you can detect that function in the lint, similar to how it detects `panic_any` / `panic_str` / etc.

File: src/test/ui/macros/unreachable-format-args.rs (hunk 4 above)
Comment (Urgau) [reply src/test/ui/macros/unreachable-format-args.rs:1]: I did the same as this test [unreachable-fmt-msg.rs](https://github.com/rust-lang/rust/blob/8196e37ac5c1ed611470faa1f73b4af2f7bc7037/src/test/ui/macros/unreachable-fmt-msg.rs#L3)

File: library/core/src/panic.rs [hunk 5]
```diff
@@ -58,6 +58,42 @@ pub macro panic_2021 {
...
+#[rustc_macro_transparency = "semitransparent"]
+pub macro unreachable_2015 {
+    () => (
+        $crate::panicking::panic("internal error: entered unreachable code")
+    ),
+    // Use `panic_str` instead of `panic_display::<&str>` for non_fmt_panic lint.
+    ($msg:expr $(,)?) => (
+        $crate::panicking::panic_str(
+            $crate::concat!("internal error: entered unreachable code: ", $msg))
+    ),
```
Comment (m-ou-se) [inline library/core/src/panic.rs:74]: Using `concat` rather than formatting with Display breaks code like this:

```rust
    let a = "hello";
    unreachable!(a);
```

Comment (Urgau) [reply library/core/src/panic.rs]: Fixed with the introduction of `unreachable_display!` and test added https://github.com/rust-lang/rust/pull/93179/commits/9264aa58d2cc38c6a497f19ef616b38c52b3fff6

**Comment by Urgau** [comment]
> > I tried to expand the current `non_fmt_panics` but I'm stuck on the fact that the argument passed to `panic_str` is the result of `$crate::concat!("internal error: entered unreachable code: ", $fmt)` and I don't know how (if even possible) to get the argument passed to `concat` or `unreachable_2015`. Any suggestions ?
> 
> Maybe you could make the second case of `unreachable_2015` expand to a function call instead: `unreachable_display(&$msg)` and then define a `fn unreachable_display(_: &impl Display) -> !;` to do the panic. Then you can detect that function in the lint, similar to how it detects `panic_any` / `panic_str` / etc.

Added in https://github.com/rust-lang/rust/pull/93179/commits/8f00aed529488bbb81fd2570167f74a9302b342b and update of `non_fmt_panic` in https://github.com/rust-lang/rust/pull/93179/commits/d8de2383e2c25bd2ad10b44b03646acf1e19247a.

File: library/core/src/panicking.rs [hunk 6]
```diff
@@ -56,6 +56,16 @@ pub const fn panic_str(expr: &str) -> ! {
     panic_display(&expr);
 }
 
+#[cfg(not(bootstrap))]
+#[inline]
+#[track_caller]
+#[lang = "unreachable_display"] // needed for `non-fmt-panics` lint
```
Comment (m-ou-se) [inline library/core/src/panicking.rs:62]: If a lang item is only used for diagnostics, it's better to use `rustc_diagnostic_item` instead.

File: library/core/src/panicking.rs [hunk 7]
```diff
@@ -56,6 +56,16 @@ pub const fn panic_str(expr: &str) -> ! {
...
+#[rustc_do_not_const_check] // hooked by const-eval
+#[rustc_const_unstable(feature = "core_panic", issue = "none")]
+pub const fn unreachable_display<T: fmt::Display>(x: &T) -> ! {
```
Comment (m-ou-se) [inline library/core/src/panicking.rs:65]: `unreachable!(arg)` has never worked in const context, and the comment `// hooked by const-eval` is not true, making `rustc_do_not_const_check` problematic. I don't think there's any need to make this function `const`.

File: library/core/src/panic.rs [hunk 8]
```diff
@@ -58,6 +58,42 @@ pub macro panic_2021 {
...
+    () => (
+        $crate::panicking::panic("internal error: entered unreachable code")
+    ),
+    ($($t:tt)+) => (
+        $crate::panicking::panic_fmt(
+            $crate::const_format_args!(
+                "internal error: entered unreachable code: {}",
+                    $crate::const_format_args!($($t)+)))
+    ),
+}
```
Comment (m-ou-se) [inline library/core/src/panic.rs:95]: Let's just use the `panic!()` macro in the second case, to follow the `todo` and `unimplemented` macros: https://github.com/rust-lang/rust/blob/a34c0797528172ede89480e3033f7a5e71ea4735/library/core/src/macros/mod.rs#L688-L691

There's no need for the `const_format_args` here. The regular `format_args` should suffice, as none of this would work in const context anyway.

File: compiler/rustc_lint/src/non_fmt_panic.rs [hunk 9]
```diff
@@ -67,7 +79,11 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
     }
 }
 
-fn check_panic<'tcx>(cx: &LateContext<'tcx>, f: &'tcx hir::Expr<'tcx>, arg: &'tcx hir::Expr<'tcx>) {
+fn check_panic_or_unreachable<'tcx>(
```
Comment (m-ou-se) [inline compiler/rustc_lint/src/non_fmt_panic.rs:82]: This function also covers `debug_assert` and `assert`, but since `check_panic_or_assert_or_debug_assert_or_unreachable` gets a bit long, I think just `check_panic` suffices, as they're all forms of panic.

File: library/core/src/panic.rs [hunk 10]
```diff
@@ -58,6 +58,42 @@ pub macro panic_2021 {
     ),
 }
 
+#[doc(hidden)]
+#[unstable(feature = "edition_panic", issue = "none", reason = "use unreachable!() instead")]
+#[allow_internal_unstable(core_panic, const_format_args)]
```
Comment (m-ou-se) [inline library/core/src/panic.rs:63]:
```suggestion
#[allow_internal_unstable(core_panic)]
```

File: compiler/rustc_lint/src/non_fmt_panic.rs [hunk 11]
```diff
@@ -52,13 +52,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
...
                     if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        let diagnostic_item = cx.tcx.get_diagnostic_name(id);
                         if matches!(
-                            cx.tcx.get_diagnostic_name(id),
+                            diagnostic_item,
                             Some(sym::core_panic_2015_macro | sym::std_panic_2015_macro)
                         ) {
-                            check_panic(cx, f, arg);
+                            check_panic_or_unreachable(cx, f, arg);
+                        } else if matches!(diagnostic_item, Some(sym::unreachable_2015_macro)) {
```
Comment (m-ou-se) [inline compiler/rustc_lint/src/non_fmt_panic.rs:64]: It's probably better to do the removal of the `AddrOf` operator based on the function that was called (`unreachable_display()`), rather than the macro (`unreachable!()`), since `unreachable!()` can also expand to `panic_fn` where this shouldn't happen.

File: compiler/rustc_lint/src/non_fmt_panic.rs [hunk 12]
```diff
@@ -52,13 +52,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                 if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_str()
+                    || Some(def_id) == cx.tcx.lang_items().unreachable_display()
                 {
                     if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        let diagnostic_item = cx.tcx.get_diagnostic_name(id);
                         if matches!(
```
Comment (m-ou-se) [inline compiler/rustc_lint/src/non_fmt_panic.rs:59]: A series of `[else] if matches!()` on the same value can just be written as a regular `match`.

**Review by m-ou-se — CHANGES_REQUESTED** [review]
Nice! Thanks for working on this!

I have a few more comments:

File: library/core/src/panicking.rs (hunk 6 above)
Comment (m-ou-se) [reply library/core/src/panicking.rs:62]: (I see that `panic_str` has the same comment. Looks like that's left over from this cleanup: #90273. Originally it wasn't only there for diagnostics. I'll clean up `panic_str` some time soon.)

**Comment by Urgau** [comment]
I've address all of review comments. Let me know if there is something else that I should modify or improve.

File: compiler/rustc_lint/src/non_fmt_panic.rs [hunk 13]
```diff
@@ -49,9 +49,11 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
     fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
         if let hir::ExprKind::Call(f, [arg]) = &expr.kind {
             if let &ty::FnDef(def_id, _) = cx.typeck_results().expr_ty(f).kind() {
+                let f_diagnostic_name = cx.tcx.get_diagnostic_name(def_id);
+
                 if Some(def_id) == cx.tcx.lang_items().begin_panic_fn()
                     || Some(def_id) == cx.tcx.lang_items().panic_fn()
-                    || Some(def_id) == cx.tcx.lang_items().panic_str()
+                    || matches!(f_diagnostic_name, Some(sym::panic_str))
```
Comment (m-ou-se) [inline compiler/rustc_lint/src/non_fmt_panic.rs:56]:
```suggestion
                    || f_diagnostic_name == Some(sym::panic_str)
```

File: compiler/rustc_lint/src/non_fmt_panic.rs [hunk 14]
```diff
@@ -61,6 +63,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
                             check_panic(cx, f, arg);
                         }
                     }
+                } else if matches!(f_diagnostic_name, Some(sym::unreachable_display)) {
```
Comment (m-ou-se) [inline compiler/rustc_lint/src/non_fmt_panic.rs:66]:
```suggestion
                } else if f_diagnostic_name == Some(sym::unreachable_display) {
```

File: compiler/rustc_lint/src/non_fmt_panic.rs [hunk 15]
```diff
@@ -61,6 +63,25 @@ impl<'tcx> LateLintPass<'tcx> for NonPanicFmt {
...
+                    if let Some(id) = f.span.ctxt().outer_expn_data().macro_def_id {
+                        if matches!(
+                            cx.tcx.get_diagnostic_name(id),
+                            Some(sym::unreachable_2015_macro)
+                        ) {
```
Comment (m-ou-se) [inline compiler/rustc_lint/src/non_fmt_panic.rs:71]: If you're only checking a single diagnostic name + def_id pair, you can use `tcx.is_diagnostic_item(sym, id)` instead.
```suggestion
                        if cx.tcx.is_diagnostic_item(sym::unreachable_2015_macro, id) {
```

**Review by m-ou-se — APPROVED** [review]
Wauw, that was fast.

Other than a few tiny nits, this looks great to me.

Can you squash the commits a bit? There's now a few commits that undo changes of previous commits in this PR, potentially making history tracking and backporting a bit complicated. Ideally this change would be backported to 1.59 (beta) too.

I'd like someone from the compiler team to also quickly look at this, since I'm not on the compiler team. Ping @oli-obk, since you were assigned by high-five.

Let's wait with r+'ing this until the FCP on https://github.com/rust-lang/rust/issues/92137 is finished. We have a few weeks left before the next release cycle.

File: src/test/ui/consts/const-eval/const_panic.stderr [hunk 16]
```diff
@@ -68,7 +68,7 @@ error[E0080]: evaluation of constant value failed
 LL | const Y_CORE: () = core::unreachable!();
    |                    ^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'internal error: entered unreachable code', $DIR/const_panic.rs:30:20
    |
-   = note: this error originates in the macro `core::unreachable` (in Nightly builds, run with -Z macro-backtrace for more info)
+   = note: this error originates in the macro `$crate::panic::unreachable_2015` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Comment (klensy) [inline src/test/ui/consts/const-eval/const_panic.stderr:71]: Previously it was `core::unreachable` and now `$crate::panic::unreachable_2015`, reference to `$crate` as rust library in error message doesn't look good.

Comment (m-ou-se) [reply src/test/ui/consts/const-eval/const_panic.stderr:71]: That's not really a problem of this PR. There are already many cases of this for other macros. Here's two examples: https://github.com/rust-lang/rust/blob/1e4067957bd5d0e12c1657e720903209ecc291dc/src/test/ui/consts/const-eval/const_panic.stderr#L7 https://github.com/rust-lang/rust/blob/1e4067957bd5d0e12c1657e720903209ecc291dc/src/test/ui/consts/const-eval/issue-43197.stderr#L42

Might be nice to improve this, but that's not something to block this PR on.

Comment (klensy) [reply src/test/ui/consts/const-eval/const_panic.stderr:71]: Oh, didn't notice, but yes, maybe this should be improved.

**Comment by matthiaskrgr** [comment]
cc https://github.com/rust-lang/rust/issues/93329

**Comment by jhpratt** [comment]
I think this might be the first time I disagree with a perf run. This should be neutral to an improvement, not a regression. I'll leave the issue filing to someone else, though.

**Comment by camelid** [comment]
@jhpratt It is weird that the changes are so large, but it's almost entirely an improvement. There's only one regression, which means it's probably spurious.

**Comment by Urgau** [comment]
Based on this comment from @m-ou-se:

> Can you squash the commits a bit? There's now a few commits that undo changes of previous commits in this PR, potentially making history tracking and backporting a bit complicated. Ideally this change would be backported to 1.59 (beta) too.

Shouldn't this PR be marked as `beta-nominated` ? or even `stable-nominated` ?

**Comment by jhpratt** [comment]
Beta seems reasonable. Not sure we need a point release for this, though.

**Comment by Urgau** [comment]
> Beta seems reasonable. Not sure we need a point release for this, though.

Well https://github.com/rust-lang/rust/pull/93394 is in the same wane as this PR and [was](https://github.com/rust-lang/rust/pull/93394#issuecomment-1029229660) `stable-nominated` and `stable-accepted`.

**Comment by m-ou-se** [comment]
> Not sure we need a point release for this, though.

As I was recently told, `stable-nominated` doesn't mean it's worth a point release. It means that if a new stable point release happens, this should be included.

I'm not sure if even in that case it should be backported, considering this change is a bit more complex than https://github.com/rust-lang/rust/pull/93394, but it's a discussion that can be had. (But we're getting close to the next release anyway, so a point release seems unlikely at this point.)

**Comment by Mark-Simulacrum** [comment]
Marking as perf-regression-triaged -- only regression appears to be due to codegen unit reshuffling, not directly related to this PR.

**Comment by alice-i-cecile** [comment]
This broke `bevy_ecs` on nightly. Here was our fix: https://github.com/bevyengine/bevy/pull/3889

**Comment by Urgau** [comment]
@alice-i-cecile This is expected, see https://github.com/rust-lang/rust/issues/92137#issuecomment-1019519285 for more context about this change and why the breaking change is considered acceptable. Also note that this change will probably be at least back-ported to beta.

**Comment by jhpratt** [comment]
I think I've said this before, but a crater run would be nice to have here. Limited to 2021 edition if it's possible (I don't think it is).

**Comment by Urgau** [comment]
> I think I've said this before, but a crater run would be nice to have here. Limited to 2021 edition if it's possible (I don't think it is).

It will be done as part of the crater run done for every new stable (it's actually done with beta but you see the point).

**Comment by jhpratt** [comment]
Yeah, I just meant to do it before so we don't have to rush a last-minute revert if it breaks too much.

**Comment by apiraino** [comment]
Beta backport accepted as per compiler team [on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-10.20.2354818.html#271444630)

Stable backport declined ([zulip discussion](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-02-10.20.2354818.html#271448198))

@rustbot label +beta-accepted -stable-nominated
